Implement type-safe API for random number generation

Xanewok opened the pull request:
Thanks for your work on the type-safe Windows CNG API! :heart: 
I found myself needing also RNG so I figured it'd be good to upload a patch here.

Tested the changes using Windows 10.0.18362.

Xanewok committed:
Implement type-safe API for random number generation

Co-authored-by: Émile Grégoire <eg@emilegregoire.ca>

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+//! Cryptographically secure random number generation.
+#![warn(missing_docs)]
```
Would be nice to not need this and document everything.

Xanewok replied:
It's a good convention to annotate with `#[deny(missing_docs)]` at the crate root for the libraries. I didn't want to enforce that here and so I forgot to remove the warning version but if you don't mind I'll put the deny version at the crate root instead.

emgre replied:
My bad, I misread the code, I thought it was an `#!allow(missing_docs)]`.

But why do we need this? I thought clippy would warn on missing docs by default, no?

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+//! Cryptographically secure random number generation.
```
A full example like I did in the other module is needed here. It helps understanding how to use it.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    /// Fills a buffer with random bytes. Specify `opts` for additional options.
+    pub fn gen_random_with_opts(
```
I think we should make this (and the associated `RandomOptions`) private. The `BCRYPT_RNG_USE_ENTROPY_IN_BUFFER` does nothing starting from Windows 8 (and should probably never be used) and `BCRYPT_USE_SYSTEM_PREFERRED_RNG` should be hidden at creation.

Xanewok replied:
While I agree that this would simplify things and that this should probably be not used as you're saying, I don't think we get to be opinionated here as just a type-safe wrapper API.

After all, what is valid to do on the lower level should be possible to do here as well.

emgre replied:
I mean `BCRYPT_USE_SYSTEM_PREFERRED_RNG` will be usable through the `RandomAlgorithmId::SystemPrefered`, so the only thing left is the `BCRYPT_RNG_USE_ENTROPY_IN_BUFFER` which is not supported on Windows 7 and lower (extended support for Windows 7 ends in 10 days).

I guess we could simply add a `gen_random_with_entropy_in_buffer` method and mark it as deprecated, but tbh, I wouldn't mind just not supporting it.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+
+/// Additional options to be used when generating random numbers.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct RandomOptions(ULONG);
```
Make this private.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+    /// *Windows 8*: Beginning with Windows 8, the RNG algorithm supports
+    /// FIPS 186-3. Keys less than or equal to 1024 bits adhere to FIPS 186-2
+    /// and keys greater than 1024 to FIPS 186-3.
+    RNG,
```
Use camel case in enums (rustfmt will probably flag this).

Xanewok replied:
`rustc` does the case linting and I wanted to retain the low level naming because it's not so intuitive (it's weird having a random generation algorithm referred to as 'RNG' where there are other options :smile: ) but you're right, I'll change that.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+    /// additional entropy for the random number. If this flag is not specified,
+    /// this function will use a random number for the entropy.
+    //
+    // *Windows 8 and later*:  This flag is ignored in Windows 8 and later.
```
Needs `///` instead of `//`.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+
+        AlgoHandle::open(id)
+            .map(RandomAlgoHandle)
+            .expect("Couldn't open random algorithm provider")
```
I'm not really comfortable with that. Expect hides a `panic!` underneath. I prefer having explicit error handling and let user deal with errors.

emgre commented on src/random.rs:
```diff
@@ -0,0 +1,325 @@
+}
+
+/// Wrapper around `AlgoHandle` that can only specify RNG algorithms.
+pub struct RandomAlgoHandle(AlgoHandle);
```
Make this private.

Xanewok replied:
This can't be private as part of a public enum variant

emgre reviewed (changes requested):
I would prefer having only two objects exposed publicly:
- `RandomAlgorithmId`: the same enum we have currently with an extra `SystemPrefered`
- `RandomNumberGenerator` (formerly `RandomAlgorithm`): a struct with two methods:
  - `fn open(id: RandomAlgorithmId) -> crate::Result<RandomNumberGenerator>`
  - `fn gen_random(&self, &mut [u8]) -> crate::Result<()>`

`RandomNumberGenerator::open` would take care of opening the algo provider or not, and also save the required flag internally.

Xanewok commented:
Makes sense, thanks for the review!

I'll modify the public API and for now apply the thumbed up suggestions tomorrow.

emgre commented:
So I spent the evening listening to some [jazz fusion](https://youtu.be/P4aFrf86_o0) and ended implementing what I had in mind for this. You can have a look on my [`random`](https://github.com/emgre/win-crypto-ng/tree/random) branch. I think the public API of this is less scattered and easier to use, while still providing all the functionalities in a safe way. Let me know what you think about it 😉

Xanewok commented:
Thanks for doing this! I left some review comments at https://github.com/emgre/win-crypto-ng/commit/38fce6ca2509204f8ee2a3c581efcd98509daaa5.

I think I prefer your `gen_random_with_entropy_in_buffer` approach in the end. Hopefully they won't add new options in the meantime :grinning:
One thing that I think we should change is the infallible `RandomNumberGenerator::system_preferred` constructor and this should be more or less good to go modulo git shenanigans :+1:

Xanewok commented:
:wave: Did you have a chance to look at the review by any chance?

Xanewok committed:
Make RandomNumberGenerator::system_preferred infallible

Xanewok committed:
Hide some already private helper handle methods

Xanewok commented:
Squashed our changes together into a single co-authored commit and added small suggestions I mentioned (e.g. simple infallible default RNG) in the review.

I think this should be ready to go now :+1:

emgre committed:
Provide rng module documentation and example.

emgre commented:
I simply added a top-level documentation and I'll merge now. Thanks!
